### PR TITLE
Board Games List

### PR DESCRIPTION
**Pull request by itsmyrice**

### CODE REVIEW

File: components/GameCard.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  return (
+    <StyledCard>
+      <StyledTitle>{game.name.value}</StyledTitle>
+      <StyledRankTitle>{game.rank}</StyledRankTitle>
+      <Image
+        src={game.thumbnail.value}
+        alt={game.name.value}
+        width={100}
+        height={100}
+      ></Image>
```
Comment (F-Kirchhoff) [inline components/GameCard.js:14]: please use a self closing tag here :
```suggestion
      />
```

File: components/GameCard.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      <p>{game.yearpublished.value}</p>
+    </StyledCard>
+  );
+}
+
+const StyledRankTitle = styled.p`
+  color: #ff8200;
+`;
+const StyledCard = styled.div`
+  width: 80%;
```
Comment (F-Kirchhoff) [inline components/GameCard.js:24]: instead of using this percentage wise padding, please give the container element of the Cards a padding or a max-width value.

File: pages/api/games/index.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
+import { parseString } from "xml2js";
+const baseApiUrl = "https://boardgamegeek.com/xmlapi2/";
+
+// convertor function from library
+const parseXMLToJSON = async (xmlData) => {
+  return new Promise((resolve, reject) => {
+    parseString(
+      xmlData,
+      { explicitArray: false, mergeAttrs: true },
+      (err, result) => {
```
Comment (F-Kirchhoff) [inline pages/api/games/index.js:10]: please use full length variable names: `error`

File: pages/api/games/index.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        if (err) {
+          console.error("Error in parseString:", err);
+          reject(err);
+        } else {
+          resolve(result);
+        }
+      }
+    );
+  });
+};
```
Comment (F-Kirchhoff) [inline pages/api/games/index.js:20]: very nice Promise Wrapping ✨

File: pages/api/games/index.js [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+// fetch xml from API and converting xml to json
+const fetcher = async (url) => {
+  const response = await fetch(url);
+  const xmlData = await response.text();
+  try {
+    const jsonData = await parseXMLToJSON(xmlData);
+    return jsonData.items.item;
```
Comment (F-Kirchhoff) [inline pages/api/games/index.js:28]: if you want to stay as generic as possible in this function I would suggest to just return the jsonData instead of the `.items.item`

File: pages/api/games/index.js [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    const dynamicUrl = request.query.endpoint;
+    const fullUrl = baseApiUrl + dynamicUrl;
+
+    try {
+      const data = await fetcher(fullUrl);
+      response.json(data);
+    } catch (error) {
+      response.status(500).json({ error: "Error to get a correct path" });
+    }
+  }
```
Comment (F-Kirchhoff) [inline pages/api/games/index.js:47]: your client will get a Timeout error if another request method than "GET" is used. You need to 
- add an "early return" to the if statement
- return a "405" error as a default behaviour after the if statement

File: pages/index.js [hunk 7]
```diff
@@ -1,7 +1,63 @@
+import useSWR from "swr";
+import GameCard from "@/components/GameCard";
+import styled from "styled-components";
+import { DYNAMIC_URL } from "../utils/dynamicURLs";
+
+const dynamicUrl = DYNAMIC_URL;
```
Comment (F-Kirchhoff) [inline pages/index.js:6]: this renaming seems to be a bit unnecessary. You can use `DYNAMIC_URL` directly in the url for useSWR.

File: styles.js [hunk 8]
```diff
@@ -1,14 +0,0 @@
...
-  *::before,
-  *::after {
-    box-sizing: border-box;
-  }
-
-  body {
-    margin: 0;
-    font-family: system-ui;
-  }
-`;
```
Comment (F-Kirchhoff) [inline styles.js:14]: why did you remove this file? this should be the GlobalStyles and they are really useful. Please add them to your project.

File: utils/dynamicURLs.js [hunk 9]
```diff
@@ -0,0 +1,8 @@
+export const DYNAMIC_URL = {  //different endpoints for fetching different categories
+    hot:{
+      boardgame: "/hot?type=boardgame",
+      rpg : "/hot?type=rpg",
+      videogame : "/hot?type=videogame"
+    }
+  }
```
Comment (F-Kirchhoff) [inline utils/dynamicURLs.js:7]: I really like it that you put this into a utils folder. Maybe you wanna do the same with the "fetcher" function and the "parseXML" function from your endpoint?

**Review by F-Kirchhoff — CHANGES_REQUESTED** [review]
Hey GameDalf Crew!
a very high quality first PR. There is almost nothing that I could find to improve your code. Almost ;) Have a look at my comments.

File: components/GameCard.js (hunk 2 above)
Comment (kkenjeres) [reply components/GameCard.js:24]: const StyledCard = styled.div`
  max-width: 80%;
  }
`;

File: pages/api/games/index.js (hunk 3 above)
Comment (kkenjeres) [reply pages/api/games/index.js:10]: const parseXMLToJSON = async (xmlData) => {
  return new Promise((resolve, reject) => {
    parseString(
      xmlData,
      { explicitArray: false, mergeAttrs: true },
      (error, result) => {
        if (error) {
          console.error("Error in parseString:", error);
          reject(error);
        } else {
          resolve(result);
        }
      }
    );
  });
};

File: utils/dynamicURLs.js (hunk 9 above)
Comment (kkenjeres) [reply utils/dynamicURLs.js:7]: we are going to change it later

File: pages/index.js (hunk 7 above)
Comment (kkenjeres) [reply pages/index.js:6]: done

File: pages/api/games/index.js (hunk 6 above)
Comment (kkenjeres) [reply pages/api/games/index.js:47]: export default async function handler(request, response) {
  if (request.method === "GET") {
    // we get endpoing first and we join dynamicUrl to the baseApiUrl
    const dynamicUrl = request.query.endpoint;
    const fullUrl = baseApiUrl + dynamicUrl;

    try {
      const data = await fetcher(fullUrl);
      response.json(data);
    } catch (error) {
      response.status(500).json({ error: "Error to get a correct path" });
    }
  } else {
    return response.status(405).json({ error: "error 405" });
  }
}

File: components/GameCard.js [hunk 10]
```diff
@@ -21,7 +21,7 @@ const StyledRankTitle = styled.p`
   color: #ff8200;
 `;
 const StyledCard = styled.div`
-  width: 80%;
+  max-width: 80%;
```
Comment (F-Kirchhoff) [inline components/GameCard.js:24]: ah please put the max-width to the "parent" element of this card, most likely the ul element. one example would be: 
```
  margin: 0 auto;
  padding: 0 2rem;
  max-width: 600px;

```

Comment (F-Kirchhoff) [reply components/GameCard.js:24]: and remove the width on this element.

File: pages/index.js [hunk 11]
```diff
@@ -3,15 +3,19 @@ import GameCard from "@/components/GameCard";
...
   const { data, error, isLoading } = useSWR(
     // fetchnig data with dynamic endpoints
-    `/api/games?endpoint=/${dynamicUrl.hot.boardgame}}`
+    `/api/games?endpoint=/${DYNAMIC_URL.hot.boardgame}}`
   );
 
-  if (error) return <div>failed to load</div>;
+  if (error)
+    return (
+      <div>
```
Comment (F-Kirchhoff) [inline pages/index.js:14]: please use a p tag here or a small tag

File: styles.js [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+  *,
+  *::before,
+  *::after {
+    box-sizing: border-box;
+  }
+  body {
+    margin: 0;
+    font-family: system-ui;
+  }
+`;
```
Comment (F-Kirchhoff) [inline styles.js:13]: you need to add the GlobalStyles in the _app.js as well, otherwise they will not have an effect.

**Review by F-Kirchhoff — CHANGES_REQUESTED** [review]
Hey,
thanks for implementing. I found some small issues still, but next round the PR will be finished.

File: components/GameCard.js (hunk 10 above)
Comment (kkenjeres) [reply components/GameCard.js:24]: const StyledUList = styled.ul`
  max-width: 80%;
  list-style: none;
  display: flex;
  flex-direction: column;
  gap: 20px;
  padding: 0;
  margin: auto;
`;

File: pages/index.js (hunk 11 above)
Comment (kkenjeres) [reply pages/index.js:14]: if (!data || isLoading) return <small>loading...</small>;

File: styles.js (hunk 12 above)
Comment (kkenjeres) [reply styles.js:13]: import "../global.css";
import { SWRConfig } from "swr";
import GlobalStyle from "../styles";
export default function App({ Component, pageProps }) {
  return (
    <>
      <GlobalStyle />

      <SWRConfig
        value={{
          fetcher: (resource, init) =>
            fetch(resource, init).then((response) => response.json()),
        }}
      >
        <Component {...pageProps} />
      </SWRConfig>
    </>
  );
}

**Comment by F-Kirchhoff** [comment]
LGTM 🚀
